C API assumes Boehm-Demers-Weiser GC is in use

In the Windows build where I happen to have disabled GC (temporarily), I noticed errors like
```
src/libexpr-c/nix_api_value.cc:4:10: fatal error: gc/gc.h: No such file or directory
    4 | #include "gc/gc.h"
      |          ^~~~~~~~~
```

--- a/src/libexpr-c/local.mk
+++ b/src/libexpr-c/local.mk
@@ -17,7 +17,7 @@ libexprc_CXXFLAGS += $(INCLUDE_libutil) $(INCLUDE_libutilc) \
 
 libexprc_LIBS = libutil libutilc libstore libstorec libexpr
 
-libexprc_LDFLAGS += -pthread
+libexprc_LDFLAGS += $(THREAD_LDFLAGS)
 
 $(eval $(call install-file-in, $(d)/nix-expr-c.pc, $(libdir)/pkgconfig, 0644))
 

--- a/src/libexpr-c/nix_api_expr.cc
+++ b/src/libexpr-c/nix_api_expr.cc
@@ -155,13 +155,13 @@ void nix_gc_now()
 }
 
 #else
-void nix_gc_incref(nix_c_context * context, const void *)
+nix_err nix_gc_incref(nix_c_context * context, const void *)
 {
     if (context)
         context->last_err_code = NIX_OK;
     return NIX_OK;
 }
-void nix_gc_decref(nix_c_context * context, const void *)
+nix_err nix_gc_decref(nix_c_context * context, const void *)
 {
     if (context)
         context->last_err_code = NIX_OK;

--- a/src/libexpr-c/nix_api_external.cc
+++ b/src/libexpr-c/nix_api_external.cc
@@ -1,7 +1,6 @@
 #include "attr-set.hh"
 #include "config.hh"
 #include "eval.hh"
-#include "gc/gc.h"
 #include "globals.hh"
 #include "value.hh"
 
@@ -16,8 +15,9 @@
 #include <nlohmann/json.hpp>
 
 #ifdef HAVE_BOEHMGC
-#define GC_INCLUDE_NEW 1
-#include "gc_cpp.h"
+# include "gc/gc.h"
+# define GC_INCLUDE_NEW 1
+# include "gc_cpp.h"
 #endif
 
 void nix_set_string_return(nix_string_return * str, const char * c)

--- a/src/libexpr-c/nix_api_value.cc
+++ b/src/libexpr-c/nix_api_value.cc
@@ -1,7 +1,6 @@
 #include "attr-set.hh"
 #include "config.hh"
 #include "eval.hh"
-#include "gc/gc.h"
 #include "globals.hh"
 #include "primops.hh"
 #include "value.hh"
@@ -13,8 +12,9 @@
 #include "nix_api_value.h"
 
 #ifdef HAVE_BOEHMGC
-#define GC_INCLUDE_NEW 1
-#include "gc_cpp.h"
+# include "gc/gc.h"
+# define GC_INCLUDE_NEW 1
+# include "gc_cpp.h"
 #endif
 
 // Helper function to throw an exception if value is null
@@ -444,7 +444,7 @@ void nix_list_builder_free(ListBuilder * list_builder)
 #if HAVE_BOEHMGC
     GC_FREE(list_builder);
 #else
-    delete bb;
+    delete list_builder;
 #endif
 }
 

--- a/src/libstore-c/local.mk
+++ b/src/libstore-c/local.mk
@@ -8,7 +8,7 @@ libstorec_SOURCES := $(wildcard $(d)/*.cc)
 
 libstorec_LIBS = libutil libstore libutilc
 
-libstorec_LDFLAGS += -pthread
+libstorec_LDFLAGS += $(THREAD_LDFLAGS)
 
 # Not just for this library itself, but also for downstream libraries using this library
 

--- a/src/libutil-c/local.mk
+++ b/src/libutil-c/local.mk
@@ -13,6 +13,6 @@ libutilc_CXXFLAGS += $(INCLUDE_libutil) $(INCLUDE_libutilc)
 
 libutilc_LIBS = libutil
 
-libutilc_LDFLAGS += -pthread
+libutilc_LDFLAGS += $(THREAD_LDFLAGS)
 
 libutilc_FORCE_INSTALL := 1

--- a/tests/unit/libstore-support/tests/nix_api_store.hh
+++ b/tests/unit/libstore-support/tests/nix_api_store.hh
@@ -2,6 +2,8 @@
 ///@file
 #include "tests/nix_api_util.hh"
 
+#include "file-system.hh"
+
 #include "nix_api_store.h"
 #include "nix_api_store_internal.h"
 
@@ -37,8 +39,18 @@ public:
 protected:
     void init_local_store()
     {
-        auto tmpl = nix::getEnv("TMPDIR").value_or("/tmp") + "/tests_nix-store.XXXXXX";
+#ifdef _WIN32
+        // no `mkdtemp` with MinGW
+        auto tmpl = nix::defaultTempDir() + "/tests_nix-store.";
+        for (size_t i = 0; true; ++i) {
+            nixDir = tmpl + std::string { i };
+            if (fs::create_directory(nixDir)) break;
+        }
+#else
+        auto tmpl = nix::defaultTempDir() + "/tests_nix-store.XXXXXX";
         nixDir = mkdtemp((char *) tmpl.c_str());
+#endif
+
         nixStoreDir = nixDir + "/my_nix_store";
 
         // Options documented in `nix help-stores`

